ui/ducks/send/send.js

```diff
@@ -0,0 +1,435 @@
+import { createSlice } from '@reduxjs/toolkit';
+import abi from 'human-standard-token-abi';
+import log from 'loglevel';
+import { addHexPrefix } from 'ethereumjs-util';
+import { multiplyCurrencies } from '../../helpers/utils/conversion-util';
+import { GAS_LIMITS } from '../../../shared/constants/gas';
+import { MIN_GAS_LIMIT_HEX } from '../../pages/send/send.constants';
+
+import {
+  addGasBuffer,
+  calcGasTotal,
+  calcTokenBalance,
+  generateTokenTransferData,
+} from '../../pages/send/send.utils';
+import {
+  getAveragePriceEstimateInHexWEI,
+  getSelectedAccount,
+  getTargetAccount,
+} from '../../selectors';
+import { estimateGas } from '../../store/actions';
+import { setCustomGasLimit } from '../gas/gas.duck';
+
+const name = 'send';
+
+export const initialState = {
+  gasButtonGroupShown: true,
+  gasLimit: null,
+  gasPrice: null,
+  gasTotal: null,
+  gasIsLoading: false,
+  tokenBalance: '0x0',
+  from: '',
+  to: '',
+  amount: '0',
+  memo: '',
+  errors: {},
+  maxModeOn: false,
+  editingTransactionId: null,
+  toNickname: '',
+  ensResolution: null,
+  ensResolutionError: '',
+};
+
+const slice = createSlice({
+  name,
+  initialState,
+  reducers: {
+    updateSendErrors: (state, action) => {
+      state.errors = { ...state.errors, ...action.payload };
+    },
+    showGasButtonGroup: (state) => {
+      state.gasButtonGroupShown = true;
+    },
+    hideGasButtonGroup: (state) => {
+      state.gasButtonGroupShown = false;
+    },
+    setGasLimit: (state, action) => {
+      state.gasLimit = action.payload;
+    },
+    setGasPrice: (state, action) => {
+      state.gasPrice = action.payload;
+    },
+    setSendTokenBalance: (state, action) => {
+      state.tokenBalance = action.payload;
+    },
+    updateSendHexData: (state, action) => {
+      state.data = action.payload;
+    },
+    updateSendTo: (state, action) => {
+      state.to = action.payload.to;
+      state.toNickname = action.payload.nickname;
+    },
+    updateSendAmount: (state, action) => {
+      state.amount = action.payload;
+    },
+    setMaxModeTo: (state, action) => {
+      state.maxModeOn = action.payload;
+    },
+    setEditingTransactionId: (state, action) => {
```

Comment:
This seems like the wrong abstraction. This should only be set when the user begins editing a transaction, right? If that's the case, this might be better represented by a "startEditingTransaction" event, or something like that. Then all of the state changes relating to editing can be made in one place, and we don't allow other modules to change the "editingTransactionId" in situations where that would lead to an invalid state.

Reply:
hard agree.